File: gameballsdk/src/main/java/com/gameball/gameball/model/helpers/RequestModelHelpers.java
package com.gameball.gameball.model.helpers;

import com.gameball.gameball.model.response.PlayerAttributes;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.util.HashMap;

public class RequestModelHelpers {
    public static HashMap<String, Object> MapFromPlayerAttributes(PlayerAttributes playerAttributes){
        Gson gson = new GsonBuilder().excludeFieldsWithoutExposeAnnotation().create();

        String jsonPlayerAttributes = gson.toJson(playerAttributes);

        HashMap<String, Object> mappedAttributes =
                gson.fromJson(jsonPlayerAttributes, new TypeToken<HashMap<String, Object>>() {}.getType());

        HashMap<String, String> additionalAttributes = playerAttributes.getAdditionalAttributes();
        if(additionalAttributes != null) {
            mappedAttributes.putAll(additionalAttributes);
        }

        HashMap<String, String> customAttributes = playerAttributes.getCustomAttributes();
        if(customAttributes != null){
            mappedAttributes.put("custom", customAttributes);
        }

        return mappedAttributes;
    }

    public static PlayerAttributes MapToPlayerAtrributes(HashMap<String, Object> PlayerAttributesHashMap){
        HashMap<String, Object> tempAttr = new HashMap<>(PlayerAttributesHashMap);

        PlayerAttributes.Builder attributesBuilder = new PlayerAttributes.Builder();

        Object attrDisplayName = tempAttr.get("displayName");
        if(attrDisplayName != null)
        {
            attributesBuilder.withEmail(attrDisplayName.toString());
            tempAttr.remove("displayName");
        }

        Object attrFirstName = tempAttr.get("firstName");
        if(attrFirstName != null)
        {
            attributesBuilder.withFirstName(attrFirstName.toString());
            tempAttr.remove("firstName");
        }

        Object attrLastName = tempAttr.get("lastName");
        if(attrLastName != null)
        {
            attributesBuilder.withLastName(attrLastName.toString());
            tempAttr.remove("lastName");
        }

        Object attrEmail = tempAttr.get("email");
        if(attrEmail != null)
        {
            attributesBuilder.withEmail(attrEmail.toString());
            tempAttr.remove("email");
        }

        Object attrGender = tempAttr.get("gender");
        if(attrGender != null)
        {
            attributesBuilder.withGender(attrGender.toString());
            tempAttr.remove("gender");
        }

        Object attrMobileNumber = tempAttr.get("mobile");
        if(attrMobileNumber != null)
        {
            attributesBuilder.withMobileNumber(attrMobileNumber.toString());
            tempAttr.remove("mobile");
        }

        Object attrDateOfBirth = tempAttr.get("dateOfBirth");
        if(attrDateOfBirth != null)
        {
            attributesBuilder.withDateOfBirth(attrDateOfBirth.toString());
            tempAttr.remove("dateOfBirth");
        }

        Object attrJoinDate = tempAttr.get("joinDate");
        if(attrJoinDate != null)
        {
            attributesBuilder.withJoinDate(attrJoinDate.toString());
            tempAttr.remove("joinDate");
        }

        HashMap<String, String> attrCustom = (HashMap<String, String>) tempAttr.get("custom");
        if(attrCustom != null){
            for(HashMap.Entry<String, String> attr: attrCustom.entrySet()){
                attributesBuilder.withCustomAttribute(attr.getKey(), attr.getValue());
            }
            tempAttr.remove("custom");
        }

        for(HashMap.Entry<String, Object> attr: tempAttr.entrySet()){
            Object attrValue = tempAttr.get(attr.getKey());
            if(attrValue != null){
                attributesBuilder.withAdditionalAttributes(attr.getKey(), attrValue.toString());
            }
        }

        return attributesBuilder.build();
    }

}

File: gameballsdk/src/main/java/com/gameball/gameball/model/helpers/RequestModelHelpers.kt
package com.gameball.gameball.model.helpers

import com.gameball.gameball.model.request.CustomerAttributes
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken

object RequestModelHelpers {
    @JvmStatic
    fun MapFromCustomerAttributes(playerAttributes: CustomerAttributes): MutableMap<String, Any> {
        val gson = GsonBuilder().excludeFieldsWithoutExposeAnnotation().create()

        val jsonPlayerAttributes = gson.toJson(playerAttributes)

        val mappedAttributes: MutableMap<String, Any> =
            gson.fromJson(jsonPlayerAttributes, object : TypeToken<MutableMap<String, Any>>() {}.type)

        val additionalAttributes = playerAttributes.additionalAttributes
        if (additionalAttributes != null) {
            mappedAttributes.putAll(additionalAttributes)
        }

        val customAttributes = playerAttributes.customAttributes
        if (customAttributes != null) {
            mappedAttributes["custom"] = customAttributes
        }

        return mappedAttributes
    }

    @JvmStatic
    fun MapToCustomerAtrributes(PlayerAttributesHashMap: Map<String, Any>): CustomerAttributes {
        val tempAttr = PlayerAttributesHashMap.toMutableMap()

        val attributesBuilder = CustomerAttributes.builder()

        val attrDisplayName = tempAttr["displayName"]
        if (attrDisplayName != null) {
            attributesBuilder.displayName(attrDisplayName.toString())
            tempAttr.remove("displayName")
        }

        val attrFirstName = tempAttr["firstName"]
        if (attrFirstName != null) {
            attributesBuilder.firstName(attrFirstName.toString())
            tempAttr.remove("firstName")
        }

        val attrLastName = tempAttr["lastName"]
        if (attrLastName != null) {
            attributesBuilder.lastName(attrLastName.toString())
            tempAttr.remove("lastName")
        }

        val attrEmail = tempAttr["email"]
        if (attrEmail != null) {
            attributesBuilder.email(attrEmail.toString())
            tempAttr.remove("email")
        }

        val attrGender = tempAttr["gender"]
        if (attrGender != null) {
            attributesBuilder.gender(attrGender.toString())
            tempAttr.remove("gender")
        }

        val attrMobileNumber = tempAttr["mobile"]
        if (attrMobileNumber != null) {
            attributesBuilder.mobile(attrMobileNumber.toString())
            tempAttr.remove("mobile")
        }

        val attrDateOfBirth = tempAttr["dateOfBirth"]
        if (attrDateOfBirth != null) {
            attributesBuilder.dateOfBirth(attrDateOfBirth.toString())
            tempAttr.remove("dateOfBirth")
        }

        val attrJoinDate = tempAttr["joinDate"]
        if (attrJoinDate != null) {
            attributesBuilder.joinDate(attrJoinDate.toString())
            tempAttr.remove("joinDate")
        }

        val attrPreferredLanguage = tempAttr["preferredLanguage"]
        if (attrPreferredLanguage != null) {
            attributesBuilder.preferredLanguage(attrPreferredLanguage.toString())
            tempAttr.remove("preferredLanguage")
        }

        val attrCustom = tempAttr["custom"] as? Map<String, String>
        if (attrCustom != null) {
            for ((key, value) in attrCustom) {
                attributesBuilder.addCustomAttribute(key, value)
            }
            tempAttr.remove("custom")
        }

        for ((key, _) in tempAttr) {
            val attrValue = tempAttr[key]
            if (attrValue != null) {
                attributesBuilder.addAdditionalAttribute(key, attrValue.toString())
            }
        }

        return attributesBuilder.build()
    }
}
